pkg/kubelet/prober/prober.go

```diff
@@ -40,6 +42,9 @@ import (
 
 const maxProbeRetries = 3
 
+// Requests cache for http probe
+var httpProbeReqCache sync.Map
```

Comment:
why a global and not part of the prober struct?

Reply:
I tried to move the cache to httpprober. However, that prevent me from reusing the probeKey struct in the prober package. While move the cache to prober struct is quite simple and straightforward.

Reply:
we have bad experience caching globally because they can create memory leaks if some objects are never deleted , and those bugs were very subtle to catch. I think we should avoid creating a cache that is global

Reply:
> I tried to move the cache to httpprober. However, that prevent me from reusing the probeKey struct in the prober package. While move the cache to prober struct is quite simple and straightforward.

Well, as caching makes sense only for HTTP probes, the logical place for it is the `httpprober` from my point of view. Reusing probeKey shouldn't stop you from at least trying. You can either move it somewhere (to avoid cyclical import) and make it public, or even duplicating in in the `httpprober`, which is also ok from my point of view.

I'd suggest to look at the possibility to cache http requests in the`httpprobe.NewRequestForHTTPGetAction` and may be renaming this method. This would allow you to cache [lifecycle probes](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/lifecycle/handlers.go#L159) as well.

Reply:
> Well, as caching makes sense only for HTTP probes, the logical place for it is the httpprober from my point of view. Reusing probeKey shouldn't stop you from at least trying. You can either move it somewhere (to avoid cyclical import) and make it public, or even duplicating in in the httpprober, which is also ok from my point of view.

I agree that the logical place for the cache should be in httpprober. If you prefer, I could implement the cache in httpprober to make a comparison.

> I'd suggest to look at the possibility to cache http requests in thehttpprobe.NewRequestForHTTPGetAction and may be renaming this method. This would allow you to cache [lifecycle probes](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/lifecycle/handlers.go#L159) as well.

I read through the code and the docs, it seems that lifecycle hooks does not  get executed as frequently as normal probes, i.e. only upon container start and termination. I guess it does't worth a cache to do that.